OpenRA.Mods.Common/Traits/Player/TimeLimitManager.cs

```diff
@@ -0,0 +1,169 @@
+				new ReadOnlyDictionary<string, string>(timelimits), TimeLimitDefault.ToString(), TimeLimitLocked);
+		}
+
+		public object Create(ActorInitializer init) { return new TimeLimitManager(init.Self, this); }
+	}
+
+	public class TimeLimitManager : INotifyTimeLimit, ITick, IWorldLoaded
+	{
+		readonly TimeLimitManagerInfo info;
+		MapOptions mapOptions;
+		LabelWidget countdownLabel;
+		CachedTransform<int, string> countdown;
+		int ticksRemaining;
+
+		public int TimeLimit;
+		public string Notification;
+
+		public TimeLimitManager(Actor self, TimeLimitManagerInfo info)
+		{
+			this.info = info;
+			Notification = info.Notification;
+
+			var tl = self.World.LobbyInfo.GlobalSettings.OptionOrDefault("timelimit", info.TimeLimitDefault.ToString());
+			if (!int.TryParse(tl, out TimeLimit))
+				TimeLimit = info.TimeLimitDefault;
+
+			// Convert from minutes to ticks
+			TimeLimit *= 60 * (1000 / self.World.Timestep);
+		}
+
+		void IWorldLoaded.WorldLoaded(World w, OpenRA.Graphics.WorldRenderer wr)
+		{
+			mapOptions = w.WorldActor.Trait<MapOptions>();
+			if (string.IsNullOrWhiteSpace(info.CountdownLabel) || string.IsNullOrWhiteSpace(info.CountdownText))
+				return;
+
+			countdownLabel = Ui.Root.GetOrNull<LabelWidget>(info.CountdownLabel);
+			if (countdownLabel != null)
+			{
+				countdown = new CachedTransform<int, string>(t =>
+					info.CountdownText.F(WidgetUtils.FormatTime(t, true, w.IsReplay ? mapOptions.GameSpeed.Timestep : w.Timestep)));
+				countdownLabel.GetText = () => countdown.Update(ticksRemaining);
+			}
+		}
+
+		void ITick.Tick(Actor self)
+		{
+			if (TimeLimit <= 0)
+				return;
+
+			var ticksPerSecond = 1000 / (self.World.IsReplay ? mapOptions.GameSpeed.Timestep : self.World.Timestep);
+			ticksRemaining = TimeLimit - self.World.WorldTick;
+
+			if (ticksRemaining == 0)
+			{
+				foreach (var ntl in self.TraitsImplementing<INotifyTimeLimit>())
+					ntl.NotifyTimerExpired(self);
+
+				foreach (var p in self.World.Players)
+					foreach (var ntl in p.PlayerActor.TraitsImplementing<INotifyTimeLimit>())
+						ntl.NotifyTimerExpired(p.PlayerActor);
+
+				return;
+			}
+
+			if (ticksRemaining < 0 || info.SkipTimeRemainingNotifications)
+				return;
+
+			foreach (var m in info.TimeLimitWarnings.Keys)
+			{
+				if (ticksRemaining == m * 60 * ticksPerSecond)
+				{
+					Game.AddChatLine(Color.White, "Battlefield Control", Notification.F(m, m > 1 ? "s" : null));
+
+					var faction = self.World.LocalPlayer == null ? null : self.World.LocalPlayer.Faction.InternalName;
+					Game.Sound.PlayNotification(self.World.Map.Rules, self.World.LocalPlayer, "Speech", info.TimeLimitWarnings[m], faction);
+				}
+			}
+		}
+
+		void INotifyTimeLimit.NotifyTimerExpired(Actor self)
+		{
+			if (countdownLabel != null)
+				countdownLabel.GetText = null;
+
+			if (!info.SkipTimerExpiredNotification)
+				Game.AddChatLine(Color.White, "Battlefield Control", "Match timer has expired.");
```

Comment:
Can we please change these to "Time limit has expired" to match the lobby option name?